### PR TITLE
a blog post about golang testable code.

### PR DESCRIPTION
**Pull request by bdong-va**
image
you can click this button to get a better readable diff.

### CODE REVIEW

File: five_tips_about_testable_golang_code.md [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+
+    return search
+}
+
+type SearchServiceInterface interface {
+    Aggregation(/* exactly same variable here*/)
+}
+```
+
+It looks almost the same, the only difference is instead of `elastic.SearchService`, we passed in an `ElasticRepositoryInterface`. For this example, the compiler would not complain about the object passed in if it has `Aggregation(/* correct type variable here*/)` method. that means we can pass a mock like this into it for testing:
```
Comment (amenon-va) [inline five_tips_about_testable_golang_code.md]: You said `ElasticRepositoryInterface` but I think you mean `SearchServiceInterface` based on what you wrote above

File: five_tips_about_testable_golang_code.md [hunk 2]
```diff
@@ -0,0 +1,171 @@
+# How to write untestable code in golang
```
Comment (amenon-va) [inline five_tips_about_testable_golang_code.md]: I would really emphasize the `UN` in untestable. Go big `UNtestable` so people get it right away

**Comment by sbreckner-va** [comment]
You could mention that the correct way to do things is to "accept interfaces and return structs" in part 2.

File: five_tips_about_testable_golang_code.md [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+According to my experience in movies, if you want to be a super bad guy, you will need to work harder than the good guys. Think about it! You need to hire evil scientists, build huge death bomb to threaten the world, a good plan and a lot of money. But be a good guy basically just need some brave hearts, you do not even have known about the structure of bomb because when choosing red/blue wire to cut, good guys never mess up.
```
Comment (aabdelrahim-va) [inline five_tips_about_testable_golang_code.md]: -build a huge death bomb, or build huge death bombs?
-But to be a good guy you just basically?
-you do not even have to know about?

File: five_tips_about_testable_golang_code.md [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+|mock a method in a mock class| Yes |Yes|
+|assert mock been called with given parameter | Yes | Yes|
+|return any result in mock method | Yes |Yes|
+|mock an attribute in a mock class | Yes | No |
+|mock a function | Yes | No|
+| patch | Yes | No |
+
+![that is how good Python is.](https://imgs.xkcd.com/comics/python.png)
+
+Mock(testify) in golang is good. Just not that *magic* as Python's. When I first learn how to use it, the thought occurred to my mind most frequently is "What, I need to do this manually?" and "If I need to do this manually, what are YOU doing, mock?"
```
Comment (sbreckner-va) [inline five_tips_about_testable_golang_code.md]: "Mock (Testify) in Golang is good, just not quite as *magical* as mocking in Python. When I first learned how to use it, the thought that occurred to me most frequently was ..."

File: five_tips_about_testable_golang_code.md [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+
+~~" I am Groot." Mock replied.~~
+
+Mock in python can automatically do two things, one is **replace current real code to mock code**, another is **verify calls happen as expected**.
+
+Mock in go testify can do the second job(verify calls) but not first one(replace real code to mock). That part has to be done manually. And manually replace have more limitation than magic replace.
+
```
Comment (sbreckner-va) [inline five_tips_about_testable_golang_code.md]: "And manual replacement has more limitations than magical replacement."

File: five_tips_about_testable_golang_code.md [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+```
+
+I removed some information to make it more clear. This function takes an `elastic.SearchService`, do something to it and return `elastic.SearchService`.
+
+Test a function like that in Python is easy. Define `mock.Aggregation` somewhere, pass this mock in, assert it be called with given parameter, then assert mock been returned. done.
+
+But we cannot do the same thing in golang.
+
+Remember, golang is a strong type language. If a function takes `elastic.SearchService` as the parameter, you cannot pass anything else. The only possible way is to create a real `elastic.SearchService` object, and do some tweaking in `Aggregation` method. It is complex, and when you seeking the solution about "how to create a real elastic.SearchService object", it is easy to get lost in chasing infinite rabbits in infinite holes. You will forget what you want at first. ~~Just like a bad marriage.~~
+
```
Comment (sbreckner-va) [inline five_tips_about_testable_golang_code.md]: "and when you start seeking out the solution for ... "

File: five_tips_about_testable_golang_code.md [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+type SearchServiceMock struct {
+    mock.Mock
+}
+(ssm *SearchServiceMock) Aggregation(/* exactly same variable here*/) {
+    args := ssm.Called(/* exactly same variable here*/)
+    return args.Get(0).(*SearchServiceMock)
+}
+```
+Then we can have all benefits of strong type language, and still keep the flexible of testability.
+
```
Comment (sbreckner-va) [inline five_tips_about_testable_golang_code.md]: "Then we can have all the benefits of a strong type language and still keep it flexible enough for testing." or something like that

File: five_tips_about_testable_golang_code.md [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+    */
+
+    return int64(listingPointScore), nil
+}
+```
+
+If same logic happens in Python, testing will be easy. `@patch` the `calculateListingScore` function and let it return `any_number_we_need, nil`, then verify the result.
+
+But in the cruel world of golang, everything is getting harder: we do not have `@patch`, and we cannot mocking function. If you call a function in somewhere, you cannot mock it at all. If you want to test a function/method included an external function call, you have to call it in real. that is not good.
+
```
Comment (sbreckner-va) [inline five_tips_about_testable_golang_code.md]: "But in the cruel world of Golang, everything is harder. We do not have '@patch' like we do in Python. If you call a function somewhere, you cannot mock it at all. If you want to test a function/method that includes an external function call, you have to call that function for real. That is not good."

File: five_tips_about_testable_golang_code.md [hunk 9]
```diff
@@ -0,0 +1,171 @@
+# How to write untestable code in golang
+According to my experience in movies, if you want to be a super bad guy, you will need to work harder than the good guys. Think about it! You need to hire evil scientists, build huge death bomb to threaten the world, a good plan and a lot of money. But be a good guy basically just need some brave hearts, you do not even have known about the structure of bomb because when choosing red/blue wire to cut, good guys never mess up.
+
+Yeah, that's not fair.
+
+Good news (?) everyone, in software developer side, mess up code is far easier than build it in a good way. If you keep following points in mind, make golang code untestable would be so easy.
```
Comment (ghost) [inline five_tips_about_testable_golang_code.md]: "on the software development side" :)

File: five_tips_about_testable_golang_code.md [hunk 10]
```diff
@@ -0,0 +1,171 @@
...
+
+Before reading this, I assume you already read [Aaron's Blog: Testify - A Toolkit for Golang Unit Tests](https://vendasta-blog.appspot.com/blog/BL-F5QKDT6H/).
+
+## 1. Always expected golang mock as powerful as Python mock
+
+This is the easiest part. If you start with something with Python, all you need is ~~love~~ do nothing extra, just write code with all Python style, you can make it untestable.
```
Comment (ghost) [inline five_tips_about_testable_golang_code.md]: "all you need is ~~love~~ to do nothing extra, just write code in Python style and you can make it untestable"

File: five_tips_about_testable_golang_code.md [hunk 11]
```diff
@@ -0,0 +1,171 @@
...
+
+|        | python mock | golang testify mock |
+|--------|-------------|-------------|
+|mock a class|  Yes|         Yes    |
+|mock a method in a mock class| Yes |Yes|
+|assert mock been called with given parameter | Yes | Yes|
+|return any result in mock method | Yes |Yes|
+|mock an attribute in a mock class | Yes | No |
+|mock a function | Yes | No|
+| patch | Yes | No |
```
Comment (ghost) [inline five_tips_about_testable_golang_code.md:24]: Sometimes I forget how much stuff you can do with markdown, this looks great

File: five_tips_about_testable_golang_code.md [hunk 12]
```diff
@@ -0,0 +1,171 @@
...
+
+![that is how good Python is.](https://imgs.xkcd.com/comics/python.png)
+
+Mock(testify) in golang is good. Just not that *magic* as Python's. When I first learn how to use it, the thought occurred to my mind most frequently is "What, I need to do this manually?" and "If I need to do this manually, what are YOU doing, mock?"
+
+~~" I am Groot." Mock replied.~~
```
Comment (ghost) [inline five_tips_about_testable_golang_code.md]: lol

File: five_tips_about_testable_golang_code.md [hunk 13]
```diff
@@ -0,0 +1,171 @@
...
+    search.Aggregation(/* put some variable here*/)
+    search.Aggregation(/*put some other variable here*/)
+
+    return search
+}
+```
+
+I removed some information to make it more clear. This function takes an `elastic.SearchService`, do something to it and return `elastic.SearchService`.
+
+Test a function like that in Python is easy. Define `mock.Aggregation` somewhere, pass this mock in, assert it be called with given parameter, then assert mock been returned. done.
```
Comment (ghost) [inline five_tips_about_testable_golang_code.md]: "Testing a function like that in Python is easy"

File: five_tips_about_testable_golang_code.md [hunk 14]
```diff
@@ -0,0 +1,171 @@
...
+    mock.Mock
+}
+(ssm *SearchServiceMock) Aggregation(/* exactly same variable here*/) {
+    args := ssm.Called(/* exactly same variable here*/)
+    return args.Get(0).(*SearchServiceMock)
+}
+```
+Then we can have all benefits of strong type language, and still keep the flexible of testability.
+
+Of course, as a bad guy who wants to make golang code untestable, you will just pass a struct into the function, nailed it.
```
Comment (ghost) [inline five_tips_about_testable_golang_code.md]: lol :+1:

File: five_tips_about_testable_golang_code.md [hunk 15]
```diff
@@ -0,0 +1,171 @@
...
+
+    return int64(listingPointScore), nil
+}
+```
+
+If same logic happens in Python, testing will be easy. `@patch` the `calculateListingScore` function and let it return `any_number_we_need, nil`, then verify the result.
+
+But in the cruel world of golang, everything is getting harder: we do not have `@patch`, and we cannot mocking function. If you call a function in somewhere, you cannot mock it at all. If you want to test a function/method included an external function call, you have to call it in real. that is not good.
+
+So we have a solution for this: put the function into a struct can be mocked by the interface.
```
Comment (ghost) [inline five_tips_about_testable_golang_code.md]: "put the function into a struct that can be mocked by the interface"

File: five_tips_about_testable_golang_code.md [hunk 16]
```diff
@@ -0,0 +1,171 @@
...
+
+    /*
+        some logic we want to test
+    */
+
+    return int64(listingPointScore), nil
+}
+```
+
+Now when we test it, we can pass a `calculatorMock` which implement with the `calculatorInterface`. Then we can do assertion or make the mock method return whatever value we want.
```
Comment (ghost) [inline five_tips_about_testable_golang_code.md]: "which implements"

File: five_tips_about_testable_golang_code.md [hunk 17]
```diff
@@ -0,0 +1,171 @@
...
+
+## 4. TDD? never heard of it.
+For Python, almost everything is testable, mock is too magic, we can always have good test coverage for all code style. In this case, I will say TDD is still good to have but not logically required for good code. You can always find a way to test your code after everything works.
+
+But for Golang, TDD have huge benefits on writing testable code. Due to the limitation of strong type code style, it is so easy to write totally worked but almost not testable code. If we do not tighten the fine string of testable in our brain, the refactoring of code to make everything tested will be painful. Not only for the developer who write it, also the one who maintain it in future. ~~and it could be you!~~
+
+If TDD is too hard for some code, just keep the testable idea in mind would still help a lot. After we write a method name and start to pass things in, there should be a tiny checklist in the corner of your brain somewhere, like:
+- Is it testable to pass this thing in?
+- Is there any function call I cannot mock here?
+- How can we reuse the mock which already existed?
```
Comment (ghost) [inline five_tips_about_testable_golang_code.md:160]: Great conclusion Bo!

**Review by ghost — COMMENTED** [review]
I really enjoyed this Bo, I like your style!

File: five_tips_about_testable_golang_code.md [hunk 18]
```diff
@@ -0,0 +1,171 @@
+# How to write UNtestable code in golang
+According to my experience in movies, if you want to be a super bad guy, you will need to work harder than the good guys. Think about it! You need to hire evil scientists, build a huge death bomb to threaten the world, a good plan and a lot of money. But be a good guy you just basically need some brave hearts, you do not even have to know about the structure of bomb because when choosing red/blue wire to cut, good guys never mess up.
```
Comment (aabdelrahim-va) [inline five_tips_about_testable_golang_code.md]: -But to be a good guy you just basically

-know about the structure of the bomb

File: five_tips_about_testable_golang_code.md [hunk 19]
```diff
@@ -0,0 +1,171 @@
...
+
+Yeah, that's not fair.
+
+Good news (?) everyone, in software developer side, mess up code is far easier than build it in a good way. If you keep following points in mind, make golang code untestable would be so easy.
```
Comment (aabdelrahim-va) [inline five_tips_about_testable_golang_code.md]: -messing up code is far easier than building it in a good way

-making golang code untestable would be so easy

File: five_tips_about_testable_golang_code.md [hunk 20]
```diff
@@ -0,0 +1,171 @@
...
+
+Mock(testify) in golang is good. Just not quite as *magical* as mocking in Python. When I first learn how to use it, the thought occurred to me most frequently was "What, I need to do this manually?" and "If I need to do this manually, what are YOU doing, mock?"
+
+~~" I am Groot." Mock replied.~~
+
+Mock in python can automatically do two things, one is **replace current real code to mock code**, another is **verify calls happen as expected**.
+
+Mock in go testify can do the second job(verify calls) but not first one(replace real code to mock). That part has to be done manually. And manually replacement have more limitations than magic replacement.
+
+We can talk about the limitations in next sections, for this section, just simply forget golang mock is not as powerful as python's mock, is a good start to write untestable code in go.
```
Comment (aabdelrahim-va) [inline five_tips_about_testable_golang_code.md]: just simply forgetting that mock is not as powerful as python's mock, is a good start to writing untestable code in go

**Pull request merged**